cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,324 @@
+ * @param {string} appName the application name
+ * @param {string} envName the environment name
+ * @param {string} certArn arn for the certificate to replicate from
+ * @param {string} envRegionAcm acm client in environment region
+ * @param {string} targetRegionAcm acm client in target region
+ * @returns {string} ARN of the replicated certificate
+ */
+const replicateCertificate = async function (
+  requestId,
+  appName,
+  envName,
+  certArn,
+  envRegionAcm,
+  targetRegionAcm
+) {
+  const { Certificate } = await envRegionAcm
+    .describeCertificate({
+      CertificateArn: certArn,
+    })
+    .promise();
+  const domainName = Certificate.DomainName;
+  const sans = Certificate.SubjectAlternativeNames;
+
+  const crypto = require("crypto");
+  return targetRegionAcm
+    .requestCertificate({
+      DomainName: domainName,
+      SubjectAlternativeNames: sans,
+      IdempotencyToken: crypto
+        .createHash("sha256")
+        .update(requestId)
+        .digest("hex")
+        .substr(0, 32),
+      ValidationMethod: "DNS",
+      Tags: [
+        {
+          Key: "copilot-application",
+          Value: appName,
+        },
+        {
+          Key: "copilot-environment",
+          Value: envName,
+        },
+      ],
+    })
+    .promise();
+};
+
+/**
+ * Deletes a certificate from AWS Certificate Manager (ACM) by its ARN.
+ * Specifically, if it is the last certificate attaching to the listener, it will also remove the CNAME records
+ * for validation in all the root, app, and env hosted zones in parallel.
+ * If the certificate does not exist, the function will return normally.
+ *
+ * @param {string} arn The certificate ARN
+ * @param {string} targetRegionAcm acm client in target region
+ */
+const deleteCertificate = async function (arn, acm) {
+  try {
+    console.log(`Waiting for certificate ${arn} to become unused`);
+
+    let inUseByResources;
+    let options;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: arn,
+        })
+        .promise();
+
+      inUseByResources = Certificate.InUseBy || [];
+      options = Certificate.DomainValidationOptions || [];
+      let ok = false;
+      for (const option of options) {
+        if (!option.ResourceRecord) {
+          ok = false;
+          break;
+        }
+        ok = true;
+      }
+      if (!ok || inUseByResources.length) {
+        // Deleting resources can be quite slow - so just sleep 30 seconds between checks.
+        await sleep(30000);
+      } else {
+        break;
+      }
+    }
+    if (inUseByResources.length) {
+      throw new Error(
+        `Certificate still in use after checking for ${maxAttempts} attempts.`
+      );
+    }
+
+    await acm
+      .deleteCertificate({
+        CertificateArn: arn,
+      })
+      .promise();
+  } catch (err) {
+    if (err.name !== "ResourceNotFoundException") {
```

Comment:
It could also be thrown by `deleteCertificate` right 🤔

Reply:
I guess it'd just be clearer to reader which calls are throwing `ResourceNotFoundException` and what errors we are handling. This huge catch block read a little confusing to me in the first glance 😂 But it could just be me.